runtime/jvm/src/main/kotlin/kotlinx/serialization/Resolving.kt

```diff
@@ -47,38 +48,76 @@ private fun mapJavaClassNameToKotlin(s: String): String = when(s) {
 }
 
 fun <E> serializerByValue(value: E, context: SerialContext? = null): KSerializer<E> {
-    val klass = (value as? Any)?.javaClass?.kotlin ?: throw SerializationException("Cannot determine class for value $value")
+    val klass =
+        (value as? Any)?.javaClass?.kotlin ?: throw SerializationException("Cannot determine class for value $value")
     return serializerByClass(klass, context)
 }
 
-fun <E> serializerBySerialDescClassname(className: String, context: SerialContext? = null): KSerializer<E> = SerialCache.lookupSerializer(className, context = context)
+fun <E> serializerBySerialDescClassname(className: String, context: SerialContext? = null): KSerializer<E> =
+    SerialCache.lookupSerializer(className, context = context)
+
+fun <E> serializerByClass(klass: KClass<*>, context: SerialContext? = null): KSerializer<E> =
+    SerialCache.lookupSerializer(mapJavaClassNameToKotlin(klass.java.canonicalName ?: ""), klass, context)
 
-fun <E> serializerByClass(klass: KClass<*>, context: SerialContext? = null): KSerializer<E> = SerialCache.lookupSerializer(mapJavaClassNameToKotlin(klass.java.canonicalName ?: ""), klass, context)
+open class TypeBase<T>
+
+inline fun <reified T> typeTokenOf(): Type {
+    val base = object : TypeBase<T>() {}
+    val superType = base::class.java.genericSuperclass!!
+    return (superType as ParameterizedType).actualTypeArguments.first()!!
+}
 
 // This method intended for static, format-agnostic resolving (e.g. in adapter factories) so context is not used here.
 @Suppress("UNCHECKED_CAST")
-fun serializerByTypeToken(type: Type): KSerializer<Any> = when(type) {
-    is Class<*> -> if (!type.isArray) {
-        serializerByClass(type.kotlin)
-    } else {
-        val eType: Class<*> = type.componentType
-        val s = serializerByTypeToken(eType)
-        ReferenceArraySerializer<Any, Any>(eType.kotlin as KClass<Any>, s) as KSerializer<Any>
-    }
-    is ParameterizedType -> {
-        val rootClass = (type.rawType as Class<*>)
-        val args = (type.actualTypeArguments)
-        when {
-            List::class.java.isAssignableFrom(rootClass) -> ArrayListSerializer(serializerByTypeToken(args[0])) as KSerializer<Any>
-            Set::class.java.isAssignableFrom(rootClass) -> HashSetSerializer(serializerByTypeToken(args[0])) as KSerializer<Any>
-            Map::class.java.isAssignableFrom(rootClass) -> HashMapSerializer(serializerByTypeToken(args[0]), serializerByTypeToken(args[1])) as KSerializer<Any>
-            Map.Entry::class.java.isAssignableFrom(rootClass) -> MapEntrySerializer(serializerByTypeToken(args[0]), serializerByTypeToken(args[1])) as KSerializer<Any>
+fun serializerByTypeToken(type: Type): KSerializer<Any> {
+    return when (type) {
```

Comment:
Since this function contains a single `return`, it can be transformed to an expression body.

Reply:
had a println there before for debugging purposes, totally forgot about it